sdk/resourcemanager/Azure.ResourceManager.Core/src/Resources/SubResource.cs

```diff
@@ -1,34 +1,89 @@
 // Licensed under the MIT License.
 
+using System;
+
 namespace Azure.ResourceManager.Core
 {
+    /// <summary>
+    /// A class representing the a sub resource of ResourceIdentifier.
+    /// </summary>
+    public class SubResource : SubResource<ResourceIdentifier>
+    {
+        /// <summary>
+        /// Initializes an empty instance of <see cref="SubResource"/> for mocking.
+        /// </summary>
+        [InitializationConstructor]
+        protected SubResource() { }
+
+        /// <summary> Initializes a new instance of <see cref="SubResource"/>. </summary>
+        /// <param name="id"> ARM resource Id. </param>
+        [SerializationConstructor]
+        protected internal SubResource(string id) : base(id) { }
+    }
+
     /// <summary>
     /// A class representing a sub-resource that contains only the ID.
     /// </summary>
     [ReferenceType]
-    public partial class SubResource
+    [System.Diagnostics.CodeAnalysis.SuppressMessage("StyleCop.CSharp.MaintainabilityRules", "SA1402:File may only contain a single type", Justification = "Types differ by type argument only")]
+    public partial class SubResource<TIdentifier> : IEquatable<SubResource<TIdentifier>>, IEquatable<string>,
+        IComparable<SubResource<TIdentifier>>, IComparable<string> where TIdentifier : ResourceIdentifier
     {
         /// <summary>
-        /// Initializes an empty instance of <see cref="SubResource"/> for mocking.
+        /// Initializes an empty instance of <see cref="SubResource{TIdentifier}"/> for mocking.
         /// </summary>
         [InitializationConstructor]
-        public SubResource()
-        {
-        }
+        protected SubResource() { }
 
-        /// <summary> Initializes a new instance of SubResource. </summary>
+        /// <summary> Initializes a new instance of <see cref="SubResource{TIdentifier}"/>. </summary>
         /// <param name="id"> ARM resource Id. </param>
         [SerializationConstructor]
         protected internal SubResource(string id)
         {
-            Id = id;
+            Id = ResourceIdentifier.Create(id) as TIdentifier;
         }
 
         /// <summary>
         /// Gets the ARM resource identifier.
         /// </summary>
         /// <value></value>
-        public virtual ResourceIdentifier Id { get; }
+        public virtual TIdentifier Id { get; }
+
+        /// <inheritdoc/>
+        public int CompareTo(string other)
+        {
+            return string.Compare(Id, other, StringComparison.InvariantCultureIgnoreCase);
+        }
+
+        /// <inheritdoc/>
+        public int CompareTo(SubResource<TIdentifier> other)
+        {
+            if (other is null)
+                return 1;
+
+            if (ReferenceEquals(this, other))
+                return 0;
+
+            return Id.CompareTo(other.Id);
+        }
+
+        /// <inheritdoc/>
+        public bool Equals(SubResource<TIdentifier> other)
+        {
+            if (Id is null)
```

Comment:
Id cannot be null here since it must be set before exiting the ctor, same with Equals